### PR TITLE
[aot] Fix encoding/decoding of types with custom modifiers.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/mono/mono/issues/8701.

### CODE REVIEW

File: mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -3118,7 +3118,20 @@ encode_type (MonoAotCompile *acfg, MonoType *t, guint8 *buf, guint8 **endbuf)
 	guint8 *p = buf;
 
 	// Change memory allocation in decode_type if you change
```
Comment (jaykrell) [inline mono/mini/aot-compiler.c]: Obsolete comment?

**Comment by marek-safar** [comment]
Could you please add a test for the fix

**Comment by vargaz** [comment]
I couldn't make a test which actually reads the data.

**Comment by vargaz** [comment]
So I can add a test but it requires csharp 7.2. Is it ok to add a test which depends on that ?

**Comment by jaykrell** [comment]
Can you write it in C# but then ildasm? That's what many of my tests are..

File: mono/mini/aot-tests.cs [hunk 2]
```diff
@@ -516,4 +516,38 @@ class GetCurrentMethodClass<T> {
 		var map2 = new Dictionary <IntPtr, WeakReference> (EqualityComparer<IntPtr>.Default);
 		return 0;
 	}
+
+	// Requires c# 7.2
+#if FALSE
```
Comment (marek-safar) [inline mono/mini/aot-tests.cs]: We support 7.2, what is failing here?

**Comment by vargaz** [comment]
mcs doesn't support it:

aot-tests.cs(542,40): error CS1519: Unexpected symbol `OP_GENERICS_LT_DECL' in class, struct, or interface member declaration
aot-tests.cs(542,42): error CS1519: Unexpected symbol `>' in class, struct, or interface member declaration
aot-tests.cs(542,44): error CS1520: Class, struct, or interface method must have a return type

**Comment by marek-safar** [comment]
Use `#if !__MonoCS__` for exclusion with mcs

**Comment by marek-safar** [comment]
@monojenkins backport 2018-04

**Comment by marek-safar** [comment]
@monojenkins backport 2018-02

**Comment by marek-safar** [comment]
@vargaz please backport manually

**Comment by lewurm** [comment]
this broke the webassembly build:
```
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/mini/aot-tests.cs(538,8): error CS8107: Feature 'readonly references' is not available in C# 7.0. Please use language version 7.2 or greater.
/mnt/jenkins/workspace/test-mono-pull-request-wasm/mono/mini/aot-tests.cs(542,21): error CS8107: Feature 'readonly references' is not available in C# 7.0. Please use language version 7.2 or greater.
Makefile:133: recipe for target 'managed/mini_tests.dll' failed
```

**Comment by marek-safar** [comment]
Fixed by #8814

**Pull request merged**